### PR TITLE
Migrate to b7a 1.0.3 and introduce basic auth for the hub service

### PR DESCRIPTION
**Pull request by MaryamZi**
$title.

To Do - persist registered callbacks vs username. - Done

### CODE REVIEW

File: subscriber/src/subscriber/subscriber.bal [hunk 1]
```diff
@@ -71,21 +76,36 @@ public function main (string secret,                // secret to send to the hub
     // check whether this version is still supported
     hr = check hc->get("/isactive/" + MY_VERSION);
     if hr.statusCode != 200 {
-        io:println("*** This version of the subscriber is no longer supported!");
-        return;
+        return error("*** This version of the subscriber is no longer supported!");
```
Comment (MaryamZi) [inline subscriber/src/subscriber/subscriber.bal:79]: @sanjiva, hope this change is OK. Updated to return an error so that the program terminates.

**Review by chamil321 — APPROVED** [review]

File: subscriber/src/subscriber/subscriber.bal (hunk 1 above)
Comment (sanjiva) [reply subscriber/src/subscriber/subscriber.bal:79]: Thanks!

File: distributor/src/distributor/subscription_filter.bal [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+        string callback = paramMap.get(HUB_CALLBACK);
+        
+        string|error decodedTopic = encoding:decodeUriComponent(topic, "UTF-8");
+        if (decodedTopic is string) {
+            topic = decodedTopic;
+        } else {
+            log:printWarn("error decoding topic, using the original form: " + topic + ". Error: " + decodedTopic.toString());
+        }
+
+        if (topic != JSON_RESULTS_TOPIC) {
```
Comment (sanjiva) [inline distributor/src/distributor/subscription_filter.bal:39]: We need to support image sending too .. so user has to subscribe to both the JSON and the image. 

See: https://github.com/ECLK/Results-Dist/issues/48

Comment (sanjiva) [reply distributor/src/distributor/subscription_filter.bal:39]: We can fix this after merging this code.

File: distributor/src/distributor/subscription_filter.bal [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+        if (result is [string, string]) {
+            [string, string][username, _] = result;
+            
+            if callbackMap.hasKey(username) {
+                error? remResult = hubVar.removeSubscription(topic, callbackMap.get(username));
+                log:printInfo("Removing existing subscription for username: " + username);
+                if (remResult is error) {
+                    log:printError("error removing existing subscription for username: " + username, remResult);
+                }
+                updateUserCallback(username, callback);
```
Comment (sanjiva) [inline distributor/src/distributor/subscription_filter.bal:76]: Add a log showing that we added a new subscription and show the callback URL

File: distributor/src/distributor/subscription_filter.bal [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+        string credential = headerValue.substring(5, headerValue.length()).trim();
+
+        var result = auth:extractUsernameAndPassword(credential);
+
+        if (result is [string, string]) {
+            [string, string][username, _] = result;
+            
+            if callbackMap.hasKey(username) {
+                error? remResult = hubVar.removeSubscription(topic, callbackMap.get(username));
+                log:printInfo("Removing existing subscription for username: " + username);
```
Comment (sanjiva) [inline distributor/src/distributor/subscription_filter.bal:72]: Show the callback URL

**Review by sanjiva — APPROVED** [review]

**Pull request merged**